server.go

```diff
@@ -71,14 +80,35 @@ func New(addr string, drv *sql.Driver, provider *jsonrpc.Client, config Config,
 	go e.Start(ctx, func(ctx context.Context, b *types.Block) (func() error, error) {
 		log.Info().Msgf("Processing block: %d", b.BlockNumber)
 		var matches []processor.MatchableContract
+		var balanceUpdates []BalanceUpdate
 
 		for _, tx := range b.Transactions {
-			// txn "type" field empty. check if call data & entry point selector are empty for now
-			// to know if txn is of deploy type
-			if tx.EntryPointSelector == "" && tx.Status != types.REJECTED.String() {
-				if m := processor.Match(ctx, tx.ContractAddress, provider); m != nil {
-					log.Info().Msgf("Matched contract at %s with %s", tx.ContractAddress, m.Type())
-					matches = append(matches, m)
+			// check if transaction emitted a transfer event
+			for _, event := range tx.Events {
+				if len(event.Keys) == 0 || event.Keys[0].Cmp(caigo.GetSelectorFromName("Transfer")) != 0 {
+					continue
+				}
+
+				// check if contract indexed
+				balanceUpdate := BalanceUpdate{
+					Event: event,
+				}
+				contract, _ := client.Contract.Query().Where(contract.IDEQ(tx.ContractAddress)).Only(ctx)
+				// index contract if not indexed
+				if contract == nil {
+					matched := processor.Match(ctx, tx.ContractAddress, provider)
+					matches = append(matches, matched)
+
+					if matched.Type() != "UNKNOWN" {
+						balanceUpdate.ContractAddress = tx.ContractAddress
+						balanceUpdate.ContractType = contract.Type.String()
+						balanceUpdates = append(balanceUpdates, balanceUpdate)
+						log.Info().Msgf("Matched contract at %s with %s", tx.ContractAddress, matched.Type())
+					}
+				} else if contract.Type.String() != "UNKNOWN" {
+					balanceUpdate.ContractAddress = tx.ContractAddress
+					balanceUpdate.ContractType = contract.Type.String()
+					balanceUpdates = append(balanceUpdates, balanceUpdate)
 				}
 			}
```

Comment:
we should handle the err here (`ent.IsNotFound`) 

```suggestion
				var match MatchableContract
				contract, err := client.Contract.Query().Where(contract.IDEQ(tx.ContractAddress)).Only(ctx)
				// index contract if not indexed
				if ent.IsNotFound(err) {
					match processor.Match(ctx, tx.ContractAddress, provider)
					matches = append(matches, match)
				} else if err != nil {
				    return nil, err
				}
				
				if match.Type() != "UNKNOWN" || contract.Type.String() != "UNKNOWN" {
					balanceUpdate.ContractAddress = tx.ContractAddress
					balanceUpdate.ContractType = contract.Type.String()
					balanceUpdates = append(balanceUpdates, balanceUpdate)
				}
			}
```

Reply:
should be able to use an enum for the contract type too